Spigot-Server-Patches/0418-Ticket-level-per-player-view-distance-API.patch

```diff
@@ -0,0 +1,525 @@
++                //Paper end
++
+                 this.g.clear();
+             }
+ 
+diff --git a/src/main/java/net/minecraft/server/EntityEnderDragon.java b/src/main/java/net/minecraft/server/EntityEnderDragon.java
+index c8c74f2b3..63f19c510 100644
+--- a/src/main/java/net/minecraft/server/EntityEnderDragon.java
++++ b/src/main/java/net/minecraft/server/EntityEnderDragon.java
+@@ -570,9 +570,9 @@ public class EntityEnderDragon extends EntityInsentient implements IMonster {
+                 // CraftBukkit start - Use relative location for far away sounds
+                 // this.world.b(1028, new BlockPosition(this), 0);
+                 // Paper start
+-                int viewDistance = ((WorldServer) this.world).spigotConfig.viewDistance * 16; // Paper - updated to use worlds actual view distance incase we have to uncomment this due to removal of player view distance API
++                //int viewDistance = ((WorldServer) this.world).spigotConfig.viewDistance * 16; // Paper - updated to use worlds actual view distance incase we have to uncomment this due to removal of player view distance API
+                 for (EntityPlayer player : ((WorldServer)world).getPlayers()) {
+-                    //final int viewDistance = player.getViewDistance(); // TODO apply view distance api patch
++                    final int viewDistance = player.getTrackViewDistance(); //Paper ticket level pp-view-dist
+                     // Paper end
+                     double deltaX = this.locX - player.locX;
+                     double deltaZ = this.locZ - player.locZ;
+diff --git a/src/main/java/net/minecraft/server/EntityHuman.java b/src/main/java/net/minecraft/server/EntityHuman.java
+index cd73cde8d..809edf46f 100644
+--- a/src/main/java/net/minecraft/server/EntityHuman.java
++++ b/src/main/java/net/minecraft/server/EntityHuman.java
+@@ -80,6 +80,29 @@ public abstract class EntityHuman extends EntityLiving {
+     public String spawnWorld = "";
+     public int oldLevel = -1;
+ 
++    //Paper start ticket level pp-view-dist
++    private int viewDistance;
++    private int trackViewDistance;
++    public int lastViewDistance;
++    public int lastTrackViewDistance;
++    public int getViewDistance() { return this.viewDistance; }
++    public int getTrackViewDistance() { return this.trackViewDistance; }
++    public void setTrackViewDistance(int i) {
++        this.lastTrackViewDistance = this.trackViewDistance;
++        this.trackViewDistance = MathHelper.clamp(i, 3, 33);
++        if(i != this.lastTrackViewDistance) {
++            if (this.tracker != null) this.tracker.updatePlayer((EntityPlayer) this);
++        }
++        //set pcm vd to highest vd to respect other processes.
++        if (this.trackViewDistance > ((WorldServer) this.world).getChunkProvider().playerChunkMap.viewDistance)
++            ((WorldServer) this.world).getChunkProvider().playerChunkMap.viewDistance = this.trackViewDistance;
```

Comment:
we should not touch this variable, it will affect the mob distance map for no good reason
the mob distance map actually needs to be changed however to reflect per-player view distance

Reply:
yes this was one of the quibbles, as per player mob spawning may need to support this, although it seems that mob spawning may fall back as they can't spawn on chunks without a suitable ticket level

Reply:
it seems we may have to completely write in support for per player view distance on the per player mob spawning maps

Reply:
you really only need to change a few lines

Reply:
added in latest commit